### PR TITLE
[crater] Normalize opaques to infer vars eagerly in AssocTypeNormalizer

### PR DESCRIPTION
**Pull request by compiler-errors**
:forgor:

r? @ghost

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 1]
```diff
@@ -548,7 +515,36 @@ impl<'a, 'b, 'tcx> TypeFolder<TyCtxt<'tcx>> for AssocTypeNormalizer<'a, 'b, 'tcx
...
+                                .selcx
+                                .infcx
+                                .register_hidden_type(
+                                    ty::OpaqueTypeKey { def_id, args: data.args },
+                                    self.cause.clone(),
+                                    self.param_env,
+                                    infer,
+                                    true,
+                                )
+                                .expect("uwu");
```
Comment (matthiaskrgr) [inline compiler/rustc_trait_selection/src/traits/project.rs:541]: owo

**Comment by compiler-errors** [comment]
@craterbot check

**Comment by lcnr** [comment]
nearly all of the unknown tests are a "concrete type differs from previous defining opaque type use" where one of the hidden types is `{type error}`. Ignoring this PR, we should not emit these errors if the opaques contain a type error.

I think that most of these errors are caused by structurally instantiating opaque types, e.g. [coral-0.9.2](https://crates.io/crates/coral/0.9.2)
```rust
pub struct Message {
    pub message: String,
    pub children: Option<Vec<Message>>,
}

impl Message {
    pub fn unroll(&self) -> impl Iterator<Item = &Message> {
        let mut messages = Vec::new();
        messages.push(self);
        if let Some(ref children) = self.children {
            for child in children {
                messages.extend(child.unroll());
                // ^ the returned iter is `opaque::<'fresh>`
            }
        }
        messages.into_iter()
    }
}
```
unless we want to look through 800 regressions to figure out how many of them are actually the "needs eager inference progress case of `recur_rpit().function()`" we'd need to also change the old solver to use semantic equality for opaques

**Comment by lcnr** [comment]
There are definitely some actual regressions here, e.g.

https://github.com/kurnevsky/esxtool/blob/88538ca7c8e068b5e38d4b386eb1bf3d5cede3d0/src/mwscript/parser.rs#L186-L191

and

https://github.com/smol-rs/async-executor/blob/4fbe23af692ac7939d3a281c709c8afd4f0dac95/benches/executor.rs#L76-L80

**Comment by compiler-errors** [comment]
ok bye bye pr

**Pull request closed without merging**